pgml chat blog (#914)
postgresml/postgresml: pgml-dashboard/content/blog/pgml-chat-a-command-line-tool-for-deploying-low-latency-knowledge-based-chatbots-part-I.md

---
author: Santi Adavani
description: "pgml-chat: A command-line tool for deploying low-latency knowledge-based chatbots: Part I"
image: https://postgresml.org/dashboard/static/images/blog/pgml_vs_hf_pinecone_query.png
image_alt: "pgml-chat: A command-line tool for deploying low-latency knowledge-based chatbots: Part I"
---

# pgml-chat: A command-line tool for deploying low-latency knowledge-based chatbots: Part I
<div class="d-flex align-items-center mb-4">
  <img width="54px" height="54px" src="/dashboard/static/images/team/santi.jpg" style="border-radius: 50%;" alt="Author" />
  <div class="ps-3 d-flex justify-content-center flex-column">
    <p class="m-0">Santi Adavani</p>
    <p class="m-0">August 17, 2023</p>
  </div>
</div>

# Introduction

> Language models like GPT-3 seem really intelligent at first, but they have a huge blindspot - no external knowledge or memory. Ask them about current events or niche topics and they just can't keep up. To be truly useful in real applications, these large language models (LLMs) need knowledge added to them somehow. The trick is getting them that knowledge fast enough to have natural conversations. Open source tools like LangChain try to help by giving language models more context and knowledge. But they end up glueing together different services into a complex patchwork. This leads to a lot of infrastructure overhead, maintenance needs, and slow response times that hurt chatbot performance. We need a better solution tailored specifically for chatbots to inject knowledge in a way that's fast, relevant and integrated.

Reviewer: ```suggestion
Language models like GPT-4 seem really intelligent at first, but they have a huge blindspot - no external knowledge or memory. Ask them about current events or niche topics and they just can't keep up. To be truly useful in real applications, these large language models (LLMs) need knowledge added to them somehow. The trick is getting them that knowledge fast enough to have natural conversations. Open source tools like LangChain and LlamaIndex try to help by giving language models more context and knowledge. But they end up glueing together different services into a complex patchwork. This leads to a lot of infrastructure overhead, maintenance needs, and slow response times that hurt chatbot performance. We need a better solution tailored specifically for chatbots to inject knowledge in a way that's fast, relevant and integrated.
```

> In the first part of this blog series, we will talk about deploying a chatbot using `pgml-chat` command line tool. In the second part, we will show how `pgml-chat` works under the hood and focus on achieving low-latencies.

Reviewer: ```suggestion
In the first part of this blog series, we will talk about deploying a chatbot using the `pgml-chat` command line tool. In the second part, we will show how `pgml-chat` works under the hood and focus on achieving low-latencies.
```

 2. Splitting these documents into smaller chunks using different splitter algorithms. This keeps each chunk within the context size limits of AI models. In addition, it allows for chunking strategies that are tailored to the file type (e.g. PDFs, HTML, .py etc.).
 3. Generating semantic embeddings for each chunk using deep learning models like SentenceTransformers. The embeddings capture conceptual meaning.
 4. Indexing the chunk embeddings for efficient similarity search during conversations.

This knowledge base setup powers the contextual understanding for your chatbot. It's compute-intensive but only needs to be peridocially updated as your domain knowledge evolves.

## 2. Connecting to Conversational AI

With its knowledge base in place, now the chatbot links to models that allow natural conversations:

1. Based on users' questions, querying the indexed chunks to rapidly pull the most relevant passages.
2. Passing those passages to a model like GPT-3 to generate conversational responses.
3. Orchestrating the query, retrieval and generation flow to enable real-time chat.

> ## 3. Evaluating and Fine-tuning chatbot

Reviewer: ```suggestion
## 3. Evaluating and Fine-tuning the chatbot
```

> Chatbot needs to be evaluated and fine-tuned before it can be deployed to the real world. This involves:

Reviewer: ```suggestion
The chatbot needs to be evaluated and fine-tuned before it can be deployed to the real world. This involves:
```
